tests/unit/test_cloudserver.py

```diff
@@ -0,0 +1,99 @@
+"""CloudServer test suite"""
+import unittest
+import json
+import mock
+import requests
+from botocore.exceptions import ClientError
+import pycognito
+import fmrest
+from fmrest.exceptions import FileMakerError, BadJSON
+
+URL = 'https://111.111.111.111'
+ACCOUNT_NAME = 'demo'
+ACCOUNT_PASS = 'demo'
+DATABASE = 'Demo'
+LAYOUT = 'Demo'
+
+def _mock_authenticate_user(_, client=None):
+    """Mock Pycognito authenticate user method. This code is from Pycognito's test suite."""
+    return {
+        "AuthenticationResult": {
+            "TokenType": "admin",
+            "IdToken": "dummy_token",
+            "AccessToken": "dummy_token",
+            "RefreshToken": "dummy_token",
+        }
+    }
+
+def _mock_verify_tokens(self, token, id_name, token_use):
+    """Mock Pycognito verify tokens method. This code is from Pycognito's test suite."""
+    if "wrong" in token:
+        raise pycognito.TokenVerificationException
+    setattr(self, id_name, token)
+
+
+class CloudServerTestCase(unittest.TestCase):
+    """CloudServer test suite.
+
+    Only put mocked requests here that don't need an actual FileMaker Server.
+    """
+
+    def setUp(self) -> None:
+
+        # disable urlib warnings as we are testing with non verified certs
+        requests.packages.urllib3.disable_warnings()
+
+        self._fms = fmrest.CloudServer(url=URL,
+                                       user=ACCOUNT_NAME,
+                                       password=ACCOUNT_PASS,
+                                       database=DATABASE,
+                                       layout=LAYOUT
+                                       )
+
+    def test_bad_cognito_credentials(self):
```

Comment:
This test (`test_bad_cognito_credentials`) still needs patching, otherwise the unit test makes an actual connection to the AWS endpoint.

Reply:
Do you think it would make sense to move this to an integration test, rather than patching it here?

Reply:
I'm not sure if the intention here was to specifially test the wrong password case or more generally the interface to pycognito/boto (as other things might be caught by the `ClientError` exception as well). A real integration test would of course be best, but is currently not really possible as we don't have a dedicated test instance available (or would require to set up something like https://github.com/spulec/moto + a fake FM Cloud that knows about `FMID` for running tests).

For now, I'd be fine with just mocking the actual requests to both cognito and FM and test a positive case (i.e. `login` returns a token when both Cognito and FM calls return data as expected), since we are currently in any case not behaving differently when the `ClientError` is thrown (i.e. we don't catch the exception/raise our own).

Reply:
Understood, thanks David! I have pushed out a new test which I think accomplishes this, `test_login`. Let me know if I'm missing anything here.